File: package/main/src/Function/once.ts
/**
 * Creates a function that is restricted to be called only once.
 * Subsequent calls return the result of the first invocation.
 *
 * @param function_ - The function to restrict.
 * @returns A function that invokes the original only on its first call.
 *
 * @example
 * ```typescript
 * const initialize = once(() => {
 *   console.log("initialized");
 *   return 42;
 * });
 * initialize(); // logs "initialized", returns 42
 * initialize(); // returns 42 (no log)
 * ```
 */
export const once = <A extends unknown[], R>(
  function_: (...arguments_: A) => R,
): ((...arguments_: A) => R) => {
  let called = false;
  let result: R;

  return function (this: unknown, ...arguments_: A): R {
    if (!called) {
      called = true;
      result = function_.apply(this, arguments_);
    }
    return result;
  };
};

File: package/umt_python/src/function/once.py
from collections.abc import Callable
from typing import Any


class OnceFunction:
    """
    A function wrapper that ensures the underlying function is called only once.
    Subsequent calls return the cached result of the first invocation.

    Attributes:
        called: Whether the function has been invoked yet.
        result: The cached result of the first invocation, or None if not yet called.
    """

    def __init__(self, fn: Callable[..., Any]) -> None:
        self._fn = fn
        self.called: bool = False
        self.result: Any = None

    def __call__(self, *args: Any, **kwargs: Any) -> Any:  # noqa: ANN401
        if not self.called:
            self.called = True
            self.result = self._fn(*args, **kwargs)
        return self.result


def once(fn: Callable[..., Any]) -> OnceFunction:
    """
    Creates a function that is restricted to be called only once.
    Subsequent calls return the result of the first invocation,
    ignoring any new arguments.

    Args:
        fn: The function to restrict.

    Returns:
        A OnceFunction that invokes the original only on its first call.

    Example:
        >>> initialized = once(lambda: 42)
        >>> initialized()
        42
        >>> initialized()
        42
        >>> initialized.called
        True
    """
    return OnceFunction(fn)
